src/librustc_mir/transform/qualify_consts.rs

```diff
@@ -35,20 +35,35 @@ use super::promote_consts::{self, Candidate, TempState};
 /// What kind of item we are in.
 #[derive(Copy, Clone, Debug, PartialEq, Eq)]
 enum Mode {
-    Const,
+    /// `static` item
     Static,
+    /// `static mut` item
     StaticMut,
+    /// `const fn`
     ConstFn,
-    Fn
+    /// `const` items and anonymous constants (e.g. in array lengths)
+    OtherConst,
+    /// other `fn`
+    NonConstFn,
+}
+
+impl Mode {
+    /// Determine whether we are running in "const context".  "const context" refers
+    /// to code type-checked according to the rules of the "const type system":
+    /// the bodies of const/static items and `const fn`.
+    #[inline]
+    fn is_const_context(self) -> bool {
```

Comment:
@eddyb if you consider `is_const_context` too ambiguous, what about `requires_const_checking`?

Reply:
That seems much clearer, yes.